packages/web-components/src/dapp-wallet-bridge/DappWalletBridge.js

```diff
@@ -0,0 +1,191 @@
+   */
+  dispatchBridgeLoadedMessage(data) {
+    const ev = new CustomEvent('bridgeReady', {
+      detail: {
+        addOffer: offerConfig =>
+          this.sendMessageToBridge({
+            type: BridgeProtocol.addOffer,
+            offerConfig,
+          }),
+        requestDappConnection: petname =>
+          this.sendMessageToBridge({
+            type: BridgeProtocol.requestDappConnection,
+            petname,
+          }),
+        isDappApproved: data.isDappApproved,
+      },
+    });
+    this.dispatchEvent(ev);
+  }
+
+  /**
+   * When the bridge is first loaded, it signals `walletBridgeLoaded` and
+   * provides a `send` function for sending messages to it. The first thing
+   * the dapp should do is check if it needs approval.
+   *
+   * @param {{ detail: any; }} ev
+   */
+  onBridgeReadyMessage(ev) {
+    const { data, send } = ev.detail;
+    assert.equal(
+      data.type,
+      BridgeProtocol.loaded,
+      X`Unexpected bridge message type ${data.type}, expected ${BridgeProtocol.loaded}`,
+    );
+    assert.string(
+      this.address,
+      X`Bridge requires an address to connect to the agoric wallet.`,
+    );
+    assert.string(
+      this.chainId,
+      X`Bridge requires a chainId to connect to the agoric wallet.`,
+    );
+    send({
+      type: BridgeProtocol.checkIfDappApproved,
+      address: this.address,
+      chainId: this.chainId,
+    });
+    this.sendMessageToBridge = send;
+  }
+
+  /**
+   * After we've checked if the dapp needs approval, signal to the rest of the
+   * dapp that the bridge is ready to use.
+   *
+   * @param {{ detail: any; }} ev
+   */
+  onCheckIfDappApprovedMessage(ev) {
+    const { data } = ev.detail;
+    this.dispatchBridgeLoadedMessage(data);
+  }
+
+  dispatchBridgeMessage(ev) {
+    const bridgeMessage = new CustomEvent('bridgeMessage', {
+      detail: ev.detail,
+    });
+    this.dispatchEvent(bridgeMessage);
+  }
+
+  /**
+   * Handle messages from the wallet bridge.
+   *
+   * @param {{ detail: any; }} ev
+   */
+  onBridgeMessage(ev) {
+    console.debug('bridge message received', ev);
+    if (!this.isBridgeReady) {
+      this.onBridgeReadyMessage(ev);
+      this.isBridgeReady = true;
+      return;
+    }
+
+    switch (ev.detail.data.type) {
+      case BridgeProtocol.checkIfDappApproved:
+        this.onCheckIfDappApprovedMessage(ev);
+        break;
+      default:
+        this.dispatchBridgeMessage(ev);
+        break;
+    }
+  }
+
+  render() {
+    /** @type {import('lit-html').TemplateResult<1> | undefined} */
+    let backend;
+
+    if (this.bridgeHref === null) {
+      // If we don't know the bridge href yet, render the locator so it can
+      // tell us.
+      backend = html`<agoric-iframe-messenger
+        src=${DEFAULT_LOCATOR_URL}
+        @message=${this.onLocateMessage}
+        @error=${this.onError}
+      ></agoric-iframe-messenger>`;
+    } else {
+      // Otherwise render the bridge.
```

Comment:
It's not quickly apparent what's different between the two conditions. Consider conditionally binding `src` and `messageHandler` and then making this template call once with those values.

then you can have `const backend`. actually you won't even need to define `backend` because it can be inlined into the div.

Reply:
done